### DISCUSSION TITLE
IntegerScalar

### DISCUSSION THREAD

**Discussion by ozgurkalan** [ibis-project/ibis#6912] [Q&A] [answered]
Hi,

when count(), we get an IntegerScalar, how to cast it to python int type?

```
f"{t.count():,}"
>>>TypeError: unsupported format string passed to IntegerScalar.__format_

```

**Comment by cpcloud** [comment, marked as answer]
@ozgurkalan Thanks for the question!

You'll need to call `to_pandas()` on the expression to trigger computation:

```
In [1]: from ibis.interactive import *

In [2]: t = ex.penguins.fetch(table_name="penguins")

In [3]: f"{t.count().to_pandas():,}"
Out[3]: '344'
```

**Reply by lostmygithubaccount** [reply]
is there any way for Ibis to make this cleaner? it's all over this file for instance: https://github.com/lostmygithubaccount/ibis-analytics/blob/main/metrics.py

(wouldn't consider this high priority but nice if it could just work)

**Reply by cpcloud** [reply]
> is there any way for Ibis to make this cleaner?

Not without triggering computation AFAICT.

**Reply by cpcloud** [reply]
I honestly can't tell what would be more surprising:

- Triggering computation, one consequence being that formatting can take a very long time.
- Not triggering computation and returning a strange error message like the one from the OP.
- Not triggering computation and printing the `repr` of the expression.

**Reply by cpcloud** [reply]
I could see the argument for implementing `__format__` like this on `Expr`:

```python
    def __format__(self, spec: str) -> str:
        return format(self.to_pandas(), spec)
```

If a user is operating in interactive mode, they'll see the output of the expression as something that looks like a value, especially if it's a scalar result. Maybe that's an argument against repring a lazy object that way, but given that's the existing behavior, implementing `__format__` in a way that makes it consistent with the interactive repr might make sense.

On the other hand, anywhere `to_pandas()` is invoked the reader of the code *knows without a doubt* that computation is going to happen and so expectations are set, and something taking a long time won't be surprising.

**Reply by lostmygithubaccount** [reply]
to explain my personal issue with `to_pandas()` here, I just don't like it conceptually -- in my case I'm converting scalars (ints and floats) to numbers for a metrics dashboard, or sometimes strings. `to_pandas()` makes me think dataframe, but as a user I really just want to pull out the one thing computed and put it in a string

the `__format__` option seems decent to me...or keeping the call explicit and having a `to_scalar()` output for this case? or do nothing and document is has to be `to_pandas()`? I don't have a strong opinion on this

**Reply by NickCrews** [reply]
@cpcloud Strongly feel like the user must call some explicit method such as `.to_pandas()`.

Neither returning the basic error, nor using `repr`, point the user towards what they probably want. I would do this:

```python
def __format__(self):
    return SomeError(f"""cannot format {self.__class__.__name__}. Execute first: replace `f'result={v}'` with `f'result={v.to_pandas()}'`""")
```
